compiler/src/dotty/tools/dotc/typer/Typer.scala

```diff
@@ -1648,8 +1665,32 @@ class Typer extends Namer with TypeAssigner with Applications with Implicits wit
 
   def typedExpr(tree: untpd.Tree, pt: Type = WildcardType)(implicit ctx: Context): Tree =
     typed(tree, pt)(ctx retractMode Mode.PatternOrType)
-  def typedType(tree: untpd.Tree, pt: Type = WildcardType)(implicit ctx: Context): Tree = // todo: retract mode between Type and Pattern?
+  def typedType(tree: untpd.Tree, pt: Type = WildcardType)(implicit ctx: Context): Tree = {
+    // todo: retract mode between Type and Pattern?
+
+    /** Check that the are not mixed Any/Phantom.Any types in `&`, `|` and type bounds,
+     *  this includes Phantom.Any of different universes.
+     */
+    def checkedTops(tree: untpd.Tree): Set[Type] = {
+      def checkedTops2(tree1: untpd.Tree, tree2: untpd.Tree, msg: => Message, pos: Position): Set[Type] = {
+        val allTops = checkedTops(tree1) union checkedTops(tree2)
+        if (allTops.size > 1)
+          ctx.error(msg, tree.pos)
+        allTops
+      }
+      tree match {
+        case TypeBoundsTree(lo, hi) =>
+          checkedTops2(lo, hi, PhantomCrossedMixedBounds(lo, hi), tree.pos)
+        case untpd.InfixOp(left, op, right) =>
+          checkedTops2(left, right, PhantomCrossedMixedBounds(left, right), tree.pos)
```

Comment:
Wrong error message. (In general I prefer if one does not immediately generate error messages in new code. It's a premature optimization. Much clearer to write the error directly as a string, then one can check whether it makes sense or not).